File: commons-parsers/src/main/java/xyz/wagyourtail/commons/parsers/impl/json5/Json5Identifier.java
package xyz.wagyourtail.commons.parsers.impl.json5;

import lombok.val;
import xyz.wagyourtail.commons.core.StringUtils;
import xyz.wagyourtail.commons.core.reader.CharReader;
import xyz.wagyourtail.commons.core.reader.StringCharReader;
import xyz.wagyourtail.commons.parsers.Data;
import xyz.wagyourtail.commons.parsers.StringData;

public class Json5Identifier extends StringData.OnlyRaw<Data.SingleContent<String>> {

    public Json5Identifier(String rawContent) {
        super(rawContent, Json5Identifier::getContentChecked);
    }

    public Json5Identifier(CharReader<?> reader) {
        super(reader, Json5Identifier::getContentChecked);
    }

    public static Json5Identifier parse(String reader) {
        StringCharReader charReader = new StringCharReader(reader);
        val content = new Json5Identifier(charReader);
        charReader.expectEOS();
        return content;
    }

    public static Json5Identifier unchecked(String reader) {
        return new Json5Identifier(reader);
    }

    public static Data.SingleContent<String> getContentChecked(CharReader<?> r) {
        StringBuilder sb = new StringBuilder();
        char first = (char) r.take();
        if (first == '\\') {
            String s = takeUnicodeEscapeSequence(r);
            if (s.length() != 1) {
                throw r.createException("Invalid identifier start: " + s);
            }
            first = s.charAt(0);
        }
        if (
                !Character.isAlphabetic(first) &&
                        first != '$' &&
                        first != '_'
        ) {
            throw r.createException("Invalid identifier start: " + first);
        }
        sb.append(first);

        while (!r.exhausted()) {
            r.mark();
            char c = (char) r.take();
            if (c == '\\') {
                String s = takeUnicodeEscapeSequence(r);
                if (s.length() != 1) {
                    throw r.createException("Invalid identifier character: " + s);
                }
                c = s.charAt(0);
            }
            if (
                    Character.isAlphabetic(c) ||
                            c == '$' ||
                            c == '_'
            ) {
                sb.append(c);
                continue;
            }
            int type = Character.getType(c);
            if (
                    type == Character.NON_SPACING_MARK ||
                            type == Character.COMBINING_SPACING_MARK ||
                            type == Character.DECIMAL_DIGIT_NUMBER ||
                            type == Character.CONNECTOR_PUNCTUATION ||
                            c == '\u200C' ||
                            c == '\u200D'
            ) {
                sb.append(c);
                continue;
            }
            r.reset();
            break;
        }

        return new SingleContent<>(sb.toString());
    }

    private static String takeUnicodeEscapeSequence(CharReader<?> reader) {
        reader.expect('u');

        StringBuilder hex = new StringBuilder(4);
        for (int i = 0; i < 4; i++) {
            hex.append(reader.expect("hex", c -> (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')));
        }

        return new String(Character.toChars(Integer.parseInt(hex.toString(), 16)));
    }

    public String getValue() {
        return StringUtils.translateEscapes(getRawContent());
    }

}

File: commons-parsers-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/parsers/impl/json5/Json5Identifier.kt
package xyz.wagyourtail.commonskt.parsers.impl.json5

import xyz.wagyourtail.commonskt.parsers.Data
import xyz.wagyourtail.commonskt.parsers.StringData
import xyz.wagyourtail.commonskt.reader.CharReader
import xyz.wagyourtail.commonskt.utils.isAlphabetic
import xyz.wagyourtail.commonskt.utils.toUnicode

class Json5Identifier(rawContent: String) : StringData.OnlyRaw<Data.SingleContent<String>>(rawContent) {

    constructor(content: SingleContent<String>) : this(content.toString())

    companion object : StringDataBuilder<Json5Identifier, SingleContent<String>> {

        override fun invoke(rawContent: CharReader<*>): Json5Identifier {
            return Json5Identifier(checkedBuildContent(rawContent))
        }

        override fun checkedBuildContent(reader: CharReader<*>): SingleContent<String> {
            return SingleContent(buildString {
                var first = reader.take()
                if (first == null) {
                    throw reader.createException("Unexpected end of input")
                }
                if (first == '\\') {
                    val s = takeUnicodeEscapeSequence(reader)
                    if (s.length != 1) {
                        throw reader.createException("Invalid identifier start: $s")
                    }
                    first = s[0]
                }
                if (!first.isAlphabetic() && first != '$' && first != '_') {
                    throw reader.createException("Invalid identifier start: $first")
                }
                append(first)

                while (!reader.exhausted()) {
                    reader.mark()
                    var c = reader.take()!!
                    if (c == '\\') {
                        val s = takeUnicodeEscapeSequence(reader)
                        if (s.length != 1) {
                            throw reader.createException("Invalid identifier character: $s")
                        }
                        c = s[0]
                    }
                    if (c.isAlphabetic() || c == '$' || c == '_') {
                        append(c)
                        continue
                    }
                    val type = c.category
                    if (
                        type == CharCategory.NON_SPACING_MARK ||
                        type == CharCategory.COMBINING_SPACING_MARK ||
                        type == CharCategory.DECIMAL_DIGIT_NUMBER ||
                        type == CharCategory.CONNECTOR_PUNCTUATION ||
                        c == '\u200C' ||
                        c == '\u200D'
                    ) {
                        append(c)
                        continue
                    }
                    reader.reset()
                    break
                }


            })
        }

        fun takeUnicodeEscapeSequence(reader: CharReader<*>): String {
            reader.expect('u')

            val hex = buildString(4) {
                for (i in 0..3) {
                    append(reader.expect("hex") { it.isDigit() || it in 'a'..'f' || it in 'A'..'F' })
                }
            }

            return hex.toInt(16).toUnicode()
        }

    }

    override fun checkedBuildContent(reader: CharReader<*>): SingleContent<String> {
        return Companion.checkedBuildContent(reader)
    }

}